### ISSUE TITLE
4 bit / 8 bit model training / inference capabilities

### ISSUE THREAD

**Issue by okpatil4u** [tracel-ai/burn#464]
### Feature description

Llama.cpp has gained traction as it enables inferring models at 2,3,4,5,6,8,16 and 32 bit precision. Would it be possible to enable inference level quantisation capabilities into burn ?

### Feature motivation

Faster inference empowers deployment at edge (either web or laptops)

**Comment by nathanielsimard** [comment]
We are currently working on a GPU backend based on [wgpu](https://wgpu.rs), quantization is on our roadmap.

**Comment by thedevleon** [comment]
Related to this, I would also love to see support for i8 , i16 , i32 , i64 quantization for inference, i.e. to run models on embedded MCUs without a dedicated FPU (i.e. esp32c3, esp32c6) with the NdArray backend.

**Comment by thedevleon** [comment]
Coming back to this after a while, seems that [quantization support](https://github.com/tracel-ai/burn/pull/1963) has been added, but NdArray [still only supports float tensors](crates/burn-ndarray/src/tensor.rs) - is there any plans to add support for integer tensors in NdArray as well to make used of the quantized model?

**Comment by laggui** [comment]
> Coming back to this after a while, seems that [quantization support](https://github.com/tracel-ai/burn/pull/1963) has been added, but NdArray [still only supports float tensors](crates/burn-ndarray/src/tensor.rs) - is there any plans to add support for integer tensors in NdArray as well to make used of the quantized model?

Actually, ndarray has `NdArrayQTensor` (currently only i8 quantization type supported):

https://github.com/tracel-ai/burn/blob/2338912dafb7f9f18c8d14b851f7c7fd6ec542c7/crates/burn-ndarray/src/backend.rs#L57

But currently all [`QTensorOps`](https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/ops/qtensor.rs) just perform dequantize -> float op -> quantize by default.

Support for specific ops will come, but ndarray might not be the first targeted backend. Everything is in place though so adding new qtensor ops should be almost as straightforward as regular tensor ops 🙂

**Comment by polarathene** [comment]
FWIW, the popular [`mistral.rs` is open towards switching from Candle to Burn](https://github.com/EricLBuehler/mistral.rs/issues/1345#issuecomment-2944419115) but a blocker raised was quantization support.

It seems like steady progress is being made, is there any roadmap / checklist that can be tracked or is this issue the primary source of information on where Burn is at for this support?

---

Just a quick reference of PRs related to quantization support in Burn/CubeCL:

- **Sep => Dec 2024:**
  - https://github.com/tracel-ai/burn/pull/1963
  - https://github.com/tracel-ai/burn/pull/2275
  - https://github.com/tracel-ai/burn/pull/2301
  - https://github.com/tracel-ai/burn/pull/2479
  - https://github.com/tracel-ai/burn/pull/2604
  - https://github.com/tracel-ai/burn/pull/2627
- **Since Feb 2025:** (_note [`burn-jit` became `burn-cubecl`](https://github.com/tracel-ai/burn/pull/2809)_)
  - https://github.com/tracel-ai/burn/pull/2849
  - https://github.com/tracel-ai/burn/pull/3042
  - https://github.com/tracel-ai/cubecl/pull/578
  - https://github.com/tracel-ai/cubecl/pull/536
  - https://github.com/tracel-ai/cubecl/pull/661
  - https://github.com/tracel-ai/burn/pull/3136#pullrequestreview-2812293080
- **Docs:**
  - [Crate docs - Quantization](https://burn.dev/docs/burn/#quantization-beta) (brief section)
  - [Burn book - Quantization](https://burn.dev/burn-book/quantization.html)
- **Related issue:**
  - https://github.com/tracel-ai/cubecl/issues/162

`mistral.rs` also has GGUF support for various model architectures implemented along with the quantized model support. I assume if that project ports to Burn that [this GGUF issue](https://github.com/tracel-ai/burn/issues/1187#issuecomment-2942047999) might be relevant, but IIRC in `mistral.rs` reading in the metadata/tensors is handled separately, so it could perhaps use existing code for that in the meantime.

**Comment by laggui** [comment]
Awesome to hear for `mistral.rs`!

Quantization is still in very early support, but we're starting to stabilize some stuff. There already exists a ton of different formats and methods so it's hard to track the whole scope (e.g., GGUF). But the primary focus will be quantization for CubeCL / Burn on GPU. We are also working towards a CubeCL MLIR backend.

**Comment by antimora** [comment]
@laggui adding GGUF support should be easy as we recently added for Safetensors. We can use Candles conversion initially and then removing the bridge later (we can do it the same for Safetensors).

**Comment by leidix** [comment]
any updates on status and planned timeline for full quantizaton support?
I'm especially curious about running static int8 PTQ models (ideally imported from onnx) on edge HW

**Comment by laggui** [comment]
Just to summarize recent changes:

We improved quantization support with persistent memory optimization in the 0.19 release.

Right now, model weights can be quantized and operations are performed with float precision signal.

Most operations are not defined yet for quantized tensors, so the values have to be dequantized. But with fusion this can be fused with pretty much any other op, so we have reduced global memory reads and writes.

Matmul is the only compute operation currently defined with explicit support for quantized inputs. More to come in the future!

The [quantization section of the 0.19 release announcement](https://burn.dev/blog/release-0.19.0/#quantization) has a bit more info.

We shared some practical numbers in [our reddit post](https://www.reddit.com/r/rust/comments/1oiexhr/burn_0190_release_quantization_distributed/).

So in short, it doesn't support full static int8 quantization. It is more like weight-only compute quantization, where activations remain in float.

Regarding ONNX import support, that's another feature entirely 😅

**Comment by antimora** [comment]
Quantization has shipped since 0.19 (block and per-tensor schemes, sub-byte values, fused dequantization), and it keeps improving. Closing this general request. The remaining gaps have their own issues (e.g. #4566, #4725).
